src/tensor/cuda/device.rs

```diff
@@ -1,19 +1,57 @@
+use crate::prelude::storage_traits::CacheSize;
 use crate::shapes::{Shape, Unit};
 use crate::tensor::cpu::{Cpu, CpuError};
-use crate::tensor::{cache::TensorCache, DeviceStorage, HasErr, NoneTape, Tensor};
+use crate::tensor::{
+    cache::{CacheStorage, TensorCache},
+    DeviceStorage, HasErr, NoneTape, Tensor,
+};
 
 use cudarc::driver::{DevicePtr, DevicePtrMut, DeviceRepr};
 use cudarc::{
     cublas::{result::CublasError, CudaBlas},
-    driver::sys::CUdeviceptr,
     driver::{CudaDevice, CudaSlice, CudaStream, DeviceSlice, DriverError},
 };
 
+use core::alloc::Layout;
 use std::{
     sync::{Arc, Mutex, MutexGuard},
     vec::Vec,
 };
 
+impl<T> CacheStorage for CudaSlice<T> {
+    type Output<T2> = CudaSlice<T2>;
+
+    fn size(&self) -> usize {
+        Layout::array::<T>(self.len()).unwrap().size()
+    }
+
+    /// Unsafely converts the elements of a CudaSlice to a new type.
+    ///
+    /// # Safety
+    ///
+    /// Assuming that `self` is a valid [CudaSlice], the main safety concern is that the output
+    /// slice should be assumed to be uninitialized
+    unsafe fn transmute_elements<T2>(self) -> Self::Output<T2> {
+        let dev = self.device();
+
+        let src_layout = Layout::new::<T>().pad_to_align();
+        let dst_layout = Layout::new::<T2>().pad_to_align();
```

Comment:
I'm not sure if it's okay to use these with cuda - I don't think they follow the same layout/alignment rules as rust does.

It seems like they are just used to get number of bytes. Maybe just use `std::mem::size_of::<>() * self.data.len()`?

Reply:
I think that cuda formats memory in much the same way as the cpu, as we are able to `dtoh` and `htod` directly from one to another, any my understanding is that this is a byte-for-byte copy of the host array. I believe that we should keep this in case we wish to arrays with structs, as this is able to compute the padded size of a type as it would appear in an array.

Reply:
Yeah byte to byte I think is fine, I'm just not sure about the sizes/alignment.

pad_to_align says:
> Creates a layout by rounding the size of this layout up to a multiple of the layout’s alignment.
Which makes it sound like it can modify the length?

As far as alignment it sounds like cuda is always aligned to at least 256, which is definitely not the case for rust:

> Any address of a variable residing in global memory or returned by one of the memory allocation routines from the driver or runtime API is always aligned to at least 256 bytes.

https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#device-memory-accesses